shared/controlflow/codeql/controlflow/Cfg.qll

```diff
@@ -1514,6 +1528,260 @@ module MakeWithSplitting<
+      }
+
+      /** Holds if this block represent a normal exit. */
+      final predicate isNormal() { normal = true }
+    }
+
+    /**
+     * An exit basic block, that is, a basic block whose last node is
+     * an exit node.
+     */
+    final class ExitBasicBlock extends BasicBlock {
+      ExitBasicBlock() { this.getLastNode() instanceof ExitNode }
+    }
+
+    private module JoinBlockPredecessors {
+      predicate hasIdAndKind(JoinPredecessorBasicBlock jbp, int id, int kind) {
+        id = idOfCfgScope(jbp.(EntryBasicBlock).getScope()) and
+        kind = 0
+        or
+        not jbp instanceof EntryBasicBlock and
+        id = idOfAstNode(jbp.getFirstNode().(AstCfgNode).getAstNode()) and
+        kind = 1
+      }
+
+      string getSplitString(JoinPredecessorBasicBlock jbp) {
+        result = jbp.getFirstNode().(AstCfgNode).getSplitsString()
+        or
+        not exists(jbp.getFirstNode().(AstCfgNode).getSplitsString()) and
+        result = ""
+      }
+    }
+
+    /**
+     * Gets the `i`th predecessor of join block `jb`, with respect to some
+     * arbitrary order.
+     */
+    cached
+    JoinPredecessorBasicBlock getJoinBlockPredecessor(JoinBasicBlock jb, int i) {
+      result =
+        rank[i + 1](JoinPredecessorBasicBlock jbp, int id, int kind |
+          jbp = jb.getAPredecessor() and JoinBlockPredecessors::hasIdAndKind(jbp, id, kind)
+        |
+          jbp order by id, kind, JoinBlockPredecessors::getSplitString(jbp)
+        )
+    }
+
+    /** A basic block with more than one predecessor. */
+    final class JoinBasicBlock extends BasicBlock {
+      JoinBasicBlock() { this.getFirstNode().isJoin() }
+
+      /**
+       * Gets the `i`th predecessor of this join block, with respect to some
+       * arbitrary order.
+       */
+      JoinPredecessorBasicBlock getJoinBlockPredecessor(int i) {
+        result = getJoinBlockPredecessor(this, i)
+      }
+    }
+
+    /** A basic block that is an immediate predecessor of a join block. */
+    final class JoinPredecessorBasicBlock extends BasicBlock {
+      JoinPredecessorBasicBlock() { this.getASuccessor() instanceof JoinBasicBlock }
+    }
+
+    /** A basic block that terminates in a condition, splitting the subsequent control flow. */
+    final class ConditionBasicBlock extends BasicBlock {
+      ConditionBasicBlock() { this.getLastNode().isCondition() }
+
+      /**
+       * Holds if basic block `succ` is immediately controlled by this basic
+       * block with conditional value `s`. That is, `succ` is an immediate
+       * successor of this block, and `succ` can only be reached from
+       * the callable entry point by going via the `s` edge out of this basic block.
+       */
+      pragma[nomagic]
+      predicate immediatelyControls(BasicBlock succ, SuccessorType s) {
+        succ = this.getASuccessor(s) and
+        forall(BasicBlock pred | pred = succ.getAPredecessor() and pred != this |
+          succ.dominates(pred)
+        )
+      }
```

Comment:
In Java, this predicate is called `dominatingEdge` and it includes `bbIDominates(this, succ)` as an additional conjunct. I _think_ that ought to be semantically identical, but it might offer a better join-order since immediate dominance joined with the successor relation yields a much stronger context for the `forall`.
Also, in this general library, it's probably worth it to expand the qldoc a bit on the relationship between the concept of _controls_ and _dominance_. In short: controls equals edge dominance. See also the qldoc for `dominatingEdge` in Java.

Reply:
Also, the concept of _dominating edge_ aka `immediatelyControls` might very well be applicable beyond `ConditionBasicBlock`s, so it should possibly be available as a top-level predicate between basic blocks.

Reply:
In particular, this concept can be put into `BasicBlocks.qll` as it's independent of `.isCondition()`.

Reply:
> I think that ought to be semantically identical

That seems true to me.

> it might offer a better join-order since immediate dominance joined with the successor relation yields a much stronger context for the `forall`.

How can I validate that? I tried looking at the evaluator logs for the predicate with and without that conjunct and didn't see any noticeable difference.

Reply:
> Also, in this general library, it's probably worth it to expand the qldoc a bit on the relationship between the concept of controls and dominance.

I've written some stuff now for the `immediatelyControls` predicate. Let me know if there's more I should add.

Reply:
> How can I validate that? I tried looking at the evaluator logs for the predicate with and without that conjunct and didn't see any noticeable difference

Look at the RA with tuple counts and compare the two versions.

Reply:
Thanks, including the extra conjunct did indeed produce fewer tuples during the calculation 👍